lib/cloud/clients.go

```diff
@@ -942,3 +1058,90 @@ func (c *TestCloudClients) GetAzureRunCommandClient(subscription string) (azure.
 func (c *TestCloudClients) Close() error {
 	return nil
 }
+
+// AWSSessionCacheKeyBuilder is a helper struct wrapper around a string builder
+// that is used to efficiently build cache keys for a given region and, optionally,
+// a chain of AWS assumed-role sessions.
+type AWSSessionCacheKeyBuilder struct {
+	sb        *strings.Builder
+	roleCount int
+}
+
+// NewAWSSessionCacheKeyBuilder constructs a new cache key builder intialized
+// with a specific AWS region.
+func NewAWSSessionCacheKeyBuilder(region string) AWSSessionCacheKeyBuilder {
+	sb := strings.Builder{}
+	_, _ = sb.WriteString(region) // infalliable.
+	return AWSSessionCacheKeyBuilder{sb: &sb}
+}
+
+// AddRole adds an AWS IAM role to the cache key builder.
+func (c *AWSSessionCacheKeyBuilder) AddRole(awsRole services.AssumeRole) {
+	// Role ARN can potentially come from user input via --db-user for some
+	// protocols.
+	// Out of paranoia for crafted user input, I build the key in a way that
+	// will avoid fetching an AWS session from cache that the user should not
+	// have access to.
+	key := fmt.Sprintf(":Role[%d]:ARN[%v]:ExternalID[%v]", c.roleCount, awsRole.RoleARN, awsRole.ExternalID)
+	_, _ = c.sb.WriteString(key) // infalliable.
+	c.roleCount++
+}
+
+func (c *AWSSessionCacheKeyBuilder) String() string {
+	return c.sb.String()
+}
+
+// checkAndSetAssumeRoles checks a chain of AWS IAM roles for obvious errors that
+// would return an error from AWS API, and eliminates superfluous external IDs
+// from the chain.
+// "obvious" errors include:
+//   - ARN fails to parse, or is not an IAM Role.
+//   - a role is in a different partition than the given region.
+//   - a role is in a different account than the next role and the next role
+//     does not have an external ID.
```

Comment:
I was partly trying to follow the advice from this page https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html

> You are in the position of assuming roles on behalf of different customers like Example Corp in our previous scenario. You should assign a unique external ID to each customer and instruct them to add the external ID to their role's trust policy. You must then ensure that you always include the correct external ID in your requests to assume roles.
> You probably already have a unique identifier for each of your customers, and this unique ID is sufficient for use as an external ID. The external ID is not a special value that you need to create explicitly, or track separately, just for this purpose.
> You should always specify the external ID in your AssumeRole API calls. In addition when a customer gives you a role ARN, test whether you can assume the role both with and without the correct external ID. If you can assume the role without the correct external ID, don't store the customer's role ARN in your system. Wait until your customer has updated the role trust policy to require the correct external ID. In this way you help your customers to do the right thing, which helps to keep both of you protected against the confused deputy problem.

Specifically the part about checking that the external assume role only works with a correct external id.

> are we enforcing when the first assume role's account differs from the caller identity

no, we aren't checking caller identity for the first role to assume. I could add that, or I can drop the enforcement altogether. The purpose of this was partly enforcement of AWS advice and partly just sanity checking the role chain and providing helpful error messages, and avoiding failed API calls

> When the customer assumes a role from another account, it's possible it's a 3rd party account that our customer does not have control.

You're saying that they might not have control, and that the 3rd party might not have an external id condition in their role's trust policy?

Reply:
> You're saying that they might not have control, and that the 3rd party might not have an external id condition in their role's trust policy?

This may not apply to db access, especially discovery roles. Teleport's customer could be assuming a role that's owned by their customers. May not be a bad thing to enforce an external ID.

Reply:
I just removed the checks altogether in cloud clients. I don't think it's worth calling STS every time to get the caller identity, which is what I'd need to do to enforce this.